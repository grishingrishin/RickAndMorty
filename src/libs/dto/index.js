export * from '@/libs/dto/cardDto.js';
export * from '@/libs/dto/filterDto.js';
