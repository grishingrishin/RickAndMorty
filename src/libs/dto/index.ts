export * from '@/libs/dto/characterDto.ts';
export * from '@/libs/dto/filterDto.ts';
