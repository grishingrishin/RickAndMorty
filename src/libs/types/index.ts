export * from '@/libs/types/dto.ts';
