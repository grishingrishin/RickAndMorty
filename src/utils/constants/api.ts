import type { Api } from '@/utils/types';

export const API_URL: Api = 'https://rickandmortyapi.com/api/character';
