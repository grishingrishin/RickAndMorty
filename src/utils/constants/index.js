export * from '@/utils/constants/api.js';
export * from '@/utils/constants/localStorage.js';
