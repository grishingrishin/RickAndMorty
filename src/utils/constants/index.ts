export * from '@/utils/constants/api.ts';
export * from '@/utils/constants/localStorage.ts';
