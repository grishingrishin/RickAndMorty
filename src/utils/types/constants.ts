export type Api = string;

export type LocalStorageFilter = string;
