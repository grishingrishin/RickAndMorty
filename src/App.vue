<script setup>
import { ref, computed, onBeforeMount, onMounted } from 'vue';

import { LOCAL_STORAGE_FILTER, API_URL } from '@/utils/constants';
import { formatStr, injectParams } from '@/utils/helpers';
import { cardDto } from '@/libs/dto';

import ScrollPagination from '@/components/ScrollPagination.vue';
import CharacterFilter from '@/components/CharacterFilter.vue';
import CharacterList from '@/components/CharacterList.vue';

const characters = ref([]);
const filtersApplied = ref(null);
const isFetchingPage = ref(false);
const curPage = ref(1);
const isLastPage = ref(false);

const filteredCharacters = computed(() => {
	if (!filtersApplied.value) {
		return [];
	}

	const filtered = characters.value.filter(character => {
		return Object.keys(filtersApplied.value).every(key => {
			return formatStr(character[key]) === formatStr(filtersApplied.value[key]);
		});
	});

	return filtered;
});

onBeforeMount(restoreFilter);
onMounted(init);

function init() {
	fetchCharacters();
}

function restoreFilter() {
	if (window.localStorage.getItem(LOCAL_STORAGE_FILTER)) {
		filtersApplied.value = JSON.parse(window.localStorage.getItem(LOCAL_STORAGE_FILTER));
	}
}

function applyFilter(params) {
	window.localStorage.setItem(LOCAL_STORAGE_FILTER, JSON.stringify(params));
	filtersApplied.value = params;
}

function clearFilter() {
	window.localStorage.removeItem(LOCAL_STORAGE_FILTER);
	filtersApplied.value = null;
}

async function fetchCharacters() {
	if (isFetchingPage.value || isLastPage.value) {
		return;
	}

	isFetchingPage.value = true;
	const endpoint = new URL(API_URL);

	injectParams(endpoint, {
		page: curPage.value,
	});

	const response = await window.fetch(endpoint);
	const data = await response.json();

	if (data.info.next) curPage.value += 1;
	else isLastPage.value = true;

	characters.value = [...characters.value, ...data.results.map(cardDto)];

	isFetchingPage.value = false;
}
</script>

<template>
	<div class="app">
		<ScrollPagination :disabled="isLastPage" @update="fetchCharacters">
			<div class="app__top">
				<CharacterFilter
					:value="filtersApplied"
					@apply="applyFilter"
					@clear="clearFilter"
				/>
			</div>
			<div class="app__body">
				<template v-if="filtersApplied">
					<template v-if="filteredCharacters.length">
						<CharacterList :characters="filteredCharacters" />
					</template>
					<template v-else>
						<div class="app__text">Ooops... characters is not't defined</div>
					</template>
				</template>
				<template v-else>
					<CharacterList :characters="characters" />
				</template>
			</div>
		</ScrollPagination>
	</div>
</template>

<style lang="scss">
.app {
	max-width: 920px;
	margin: 0 auto;
	padding: 14px;

	&__top {
		margin-bottom: 14px;
	}
}
</style>
