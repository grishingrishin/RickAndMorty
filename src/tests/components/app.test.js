import { describe, test, expect, vi, beforeEach, afterEach } from 'vitest';
import { mount, flushPromises } from '@vue/test-utils';

import { LOCAL_STORAGE_FILTER } from '@/utils/constants';
import { getMockCharacters } from '@/tests/libs';
import { genEvent } from '@/tests/utils';

import App from '@/App.vue';

beforeEach(() => {
	vi.spyOn(global, 'fetch').mockImplementation(fetchSpy(getMockCharacters()));

	function fetchSpy(data) {
		return function (_url) {
			return new Promise(resolve => {
				resolve({
					json: () =>
						Promise.resolve({
							...data,
						}),
				});
			});
		};
	}
});

afterEach(async () => {
	await vi.resetAllMocks();
});

// prettier-ignore
describe('App', () => {
	test('Mounted component pure component', () => {
		expect(App).toBeTruthy();

		const wrapper = mount(App);

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Mounted component with characters', async () => {
		expect(App).toBeTruthy();

		const wrapper = mount(App);
		await flushPromises();

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Fetch characters after scroll', async () => {
		expect(App).toBeTruthy();

		const wrapper = mount(App);
		await flushPromises();

		genEvent('scroll');
		await flushPromises();

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Restore previously applied filters (with available values)', async () => {
		const stubValues = {
			status: 'alive',
			gender: 'female',
		};

		expect(App).toBeTruthy();

		window.localStorage.setItem(LOCAL_STORAGE_FILTER, JSON.stringify(stubValues));

		const wrapper = mount(App);
		await flushPromises();

		expect(wrapper.html()).toMatchSnapshot();

		window.localStorage.removeItem(LOCAL_STORAGE_FILTER);
	});

	test('Restore previously applied filters (with unavaliable values)', async () => {
		const stubValues = {
			status: 'foo',
			gender: 'bar',
		};

		expect(App).toBeTruthy();

		window.localStorage.setItem(LOCAL_STORAGE_FILTER, JSON.stringify(stubValues));

		const wrapper = mount(App);
		await flushPromises();

		expect(wrapper.html()).toMatchSnapshot();

		window.localStorage.removeItem(LOCAL_STORAGE_FILTER);
	});

	test('Clear previously applied filters', async () => {
		const stubValues = {
			status: 'alive',
			gender: 'female',
		};

		expect(App).toBeTruthy();

		window.localStorage.setItem(LOCAL_STORAGE_FILTER, JSON.stringify(stubValues));

		const wrapper = mount(App);
		await flushPromises();

		expect(wrapper.html()).toMatchSnapshot();

		await wrapper
			.find('[data-filter=clear]')
			.trigger('click');

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Applying selected filters (with characters)', async () => {
		const stubValues = {
			status: 'alive',
			gender: 'female',
		};

		expect(App).toBeTruthy();

		const wrapper = mount(App);
		await flushPromises();

		expect(wrapper.html()).toMatchSnapshot();

		await wrapper
			.find('[data-filter=status]')
			.setValue(stubValues.status);
		await wrapper
			.find('[data-filter=gender]')
			.setValue(stubValues.gender);
		await wrapper
			.find('[data-filter=apply]')
			.trigger('click');

		expect(wrapper.html()).toMatchSnapshot();
	});

	// FIXME: Doesn't reset previos state
	test('Applying selected filters (without characters)', async () => {
		const stubValues = {
			status: 'dead',
			gender: 'male',
		};

		expect(App).toBeTruthy();

		const wrapper = mount(App);
		await flushPromises();

		expect(wrapper.html()).toMatchSnapshot();

		await wrapper
			.find('[data-filter=status]')
			.setValue(stubValues.status);
		await wrapper
			.find('[data-filter=gender]')
			.setValue(stubValues.gender);
		await wrapper
			.find('[data-filter=apply]')
			.trigger('click');

		expect(wrapper.html()).toMatchSnapshot();
	});
});
