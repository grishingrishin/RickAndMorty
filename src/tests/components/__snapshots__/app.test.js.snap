// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Applying selected filters (with characters) 1`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <!--v-if-->
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <ul class="character-list__grid">
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/1.jpeg" alt="Name: (Rick Sanchez)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Rick Sanchez</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/2.jpeg" alt="Name: (Morty Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Morty Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1,2,3</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/5.jpeg" alt="Name: (Jerry Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Jerry Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
      </ul>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Applying selected filters (with characters) 2`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__actions">
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="apply"> apply </button></div>
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="clear"> clear </button></div>
      </div>
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <ul class="character-list__grid">
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
      </ul>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Applying selected filters (without characters) 1`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__actions">
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="apply"> apply </button></div>
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="clear"> clear </button></div>
      </div>
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <ul class="character-list__grid">
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
      </ul>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Applying selected filters (without characters) 2`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__actions">
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="apply"> apply </button></div>
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="clear"> clear </button></div>
      </div>
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <!--v-if-->
      <div class="character-list__append"> Ooops... characters is not't defined </div>
    </div>
  </div>
</div>"
`;

exports[`App > Clear previously applied filters 1`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__actions">
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="apply"> apply </button></div>
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="clear"> clear </button></div>
      </div>
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <ul class="character-list__grid">
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
      </ul>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Clear previously applied filters 2`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <!--v-if-->
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <ul class="character-list__grid">
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/1.jpeg" alt="Name: (Rick Sanchez)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Rick Sanchez</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/2.jpeg" alt="Name: (Morty Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Morty Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1,2,3</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/5.jpeg" alt="Name: (Jerry Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Jerry Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
      </ul>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Fetch characters after scroll 1`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <!--v-if-->
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <ul class="character-list__grid">
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/1.jpeg" alt="Name: (Rick Sanchez)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Rick Sanchez</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/2.jpeg" alt="Name: (Morty Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Morty Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1,2,3</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/5.jpeg" alt="Name: (Jerry Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Jerry Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/1.jpeg" alt="Name: (Rick Sanchez)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Rick Sanchez</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/2.jpeg" alt="Name: (Morty Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Morty Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1,2,3</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/5.jpeg" alt="Name: (Jerry Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Jerry Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
      </ul>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Mounted component pure component 1`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <!--v-if-->
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <!--v-if-->
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Mounted component with characters 1`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <!--v-if-->
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <ul class="character-list__grid">
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/1.jpeg" alt="Name: (Rick Sanchez)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Rick Sanchez</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/2.jpeg" alt="Name: (Morty Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Morty Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 1,2,3</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/5.jpeg" alt="Name: (Jerry Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Jerry Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
      </ul>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Restore previously applied filters (with available values) 1`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__actions">
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="apply"> apply </button></div>
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="clear"> clear </button></div>
      </div>
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <ul class="character-list__grid">
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/3.jpeg" alt="Name: (Summer Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Summer Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
        <li class="character-list__cell">
          <div class="character-card">
            <div class="character-card__figure"><img src="https://rickandmortyapi.com/api/character/avatar/4.jpeg" alt="Name: (Beth Smith)" class="character-card__img"></div>
            <div class="character-card__body">
              <div class="character-card__name">Beth Smith</div>
              <div class="character-card__status">Alive</div>
              <div class="character-card__episodes">Episodes: 6,7,8</div>
            </div>
          </div>
        </li>
      </ul>
      <!--v-if-->
    </div>
  </div>
</div>"
`;

exports[`App > Restore previously applied filters (with unavaliable values) 1`] = `
"<div class="app">
  <div class="app__top">
    <div class="character-filter">
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="status">
          <option value="">status</option>
          <option value="alive">alive</option>
          <option value="dead">dead</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__filter"><select class="character-filter__select" data-filter="gender">
          <option value="">gender</option>
          <option value="female">female</option>
          <option value="male">male</option>
          <option value="genderless">genderless</option>
          <option value="unknown">unknown</option>
        </select></div>
      <div class="character-filter__actions">
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="apply"> apply </button></div>
        <div class="character-filter__action"><button type="button" class="character-filter__btn" data-filter="clear"> clear </button></div>
      </div>
    </div>
  </div>
  <div class="app__body">
    <div class="character-list">
      <!--v-if-->
      <div class="character-list__append"> Ooops... characters is not't defined </div>
    </div>
  </div>
</div>"
`;
